modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java

```diff
@@ -0,0 +1,388 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.cluster.ClusterNode;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.GridTopic;
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.internal.managers.encryption.GridEncryptionManager;
+import org.apache.ignite.internal.processors.affinity.AffinityTopologyVersion;
+import org.apache.ignite.internal.util.GridConcurrentHashSet;
+import org.apache.ignite.internal.util.future.GridFutureAdapter;
+import org.apache.ignite.internal.util.typedef.CI3;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.events.EventType.EVT_NODE_FAILED;
+import static org.apache.ignite.events.EventType.EVT_NODE_LEFT;
+import static org.apache.ignite.internal.managers.communication.GridIoPolicy.SYSTEM_POOL;
+
+/**
+ * Distributed process is a cluster-wide process that accumulates single nodes results to finish itself.
+ * <p>
+ * The process consists of the following phases:
+ * <ol>
+ *  <li>Initial request starts process. (Sent via discovery)</li>
+ *  <li>Each server node process an initial request and send the single node result to the coordinator. (Sent via
+ *  communication)</li>
+ *  <li>The coordinator accumulate all single nodes results and finish process. (Sent via discovery)</li>
+ * </ol>
+ * <p>
+ * Several processes of one type can be started at the same time.
+ *
+ * @param <I> Request type.
+ * @param <R> Result type.
+ */
+public class DistributedProcess<I extends Serializable, R extends Serializable> {
+    /** Kernal context. */
+    private final DistributedProcessType type;
+
+    /** Active processes. */
+    private final ConcurrentHashMap<UUID, Process> processes = new ConcurrentHashMap<>(1);
+
+    /** Synchronization mutex for coordinator initializing and the remaining collection operations. */
+    private final Object mux = new Object();
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /**
+     * @param ctx Kernal context.
+     * @param type Process type.
+     * @param exec Execute action and returns future with the single node result to send to the coordinator.
+     * @param finish Finish process closure. Called on each node when all single nodes results received.
+     */
+    public DistributedProcess(GridKernalContext ctx, DistributedProcessType type,
+        Function<I, IgniteInternalFuture<R>> exec,
+        CI3<UUID, Map<UUID, R>, Map<UUID, Exception>> finish) {
+        this.ctx = ctx;
+        this.type = type;
+
+        log = ctx.log(getClass());
+
+        ctx.discovery().setCustomEventListener(InitMessage.class, (topVer, snd, msg) -> {
```

Comment:
Events in the `ctx.event().addDiscoveryEventListener` will be processed by `disco-event-worker`. Events in the `ctx.discovery().setCustomEventListener` will be processed by `disco-notifier-worker`.

We should be able to reject a cache start during a master key change to encrypt group keys using the relevant master key. 

For these purposes I have implemented the `isMasterKeyChangeInProgress` flag. A node checks this flag when process a cache start. So it should be updated in the `disco-notifier-worker` as cache start processed there too.

Otherwise, nodes can inconsistency reject starting the cache and it will lead to a hang of the exchange.